ruby/object_oriented_programming_basics/managing_ruby_projects.md

```diff
@@ -0,0 +1,227 @@
+
+Create a new Ruby file `main.rb` in a directory called `colorful`:
+```rb
+require 'colorize'
+
+puts 'Red goes faster!'.colorize(:red)
+
+puts "I'm blue da ba dee da ba di!".colorize(:blue)
+
+puts "It ain't easy bein' green...".colorize(:green)
+```
+You're probably itching to see all those colours, so run your file with `ruby foo.rb` to see them... or rather, a LoadError. Right - you need to install that gem first! Do that with `gem install colorize` and you'll see RubyGems in action. Your system now has access to the `Colorize` gem!
+
+Your system now has access to the `Colorize` gem - what about others that would like to use your code? Yeah, they would need to `gem install` it as well, no big deal. But what if you have dozens of gems? How do you make sure that the versions you use are the versions others download? This sounds rather tedious. Enter: Bundler. Another gem, part of RubyGems but released independently. Bundler allows you to declare what gems your project needs - down to their version. As for others, Bundler allows them to take that declaration, a simple file called `Gemfile` and use it to install those gems in a quick `bundle install`.  Since gem installs are global, you need a way to run only those particular gem versions that are declared in the `Gemfile`. You can do that by using `bundle exec` followed by a command that you want to execute - most likely `bundle exec ruby foo.rb`.
+
+Let's make sure whoever wants to use our script can do so:
+```bash
+bundle init
+bundle add colorize
+```
+This has created two files: `Gemfile` and `Gemfile.lock`. Let's take a look into both:
+```rb
+# Gemfile
+
+# frozen_string_literal: true
+
+source "https://rubygems.org"
+
+# gem "rails"
+
+gem "colorize", "~> 1.1"
+
+# Gemfile.lock
+GEM
+  remote: https://rubygems.org/
+  specs:
+    colorize (1.1.0)
+
+PLATFORMS
+  ruby
+  x86_64-linux
+
+DEPENDENCIES
+  colorize (~> 1.1)
+
+BUNDLED WITH
+   2.5.4
+```
+There's not much in those but as you can see, the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - basically, the first number is the major version, second is the minor version and a third, if it exists is the patch number. Major versions can break things from previous versions - think changing method names for example. Minor versions can add and change things but they can't break anything. Patches happen when you introduce bug fixes that don't break anything. So, if people behind a gem maintain it in line with semantic versioning, you can rely on this pessimistic constraint never letting your project have a gem version that could potentially break your app - it is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.
```

Comment:
The problem I found is that if I want to explain the syntax, there's no way of avoiding the semver, because it depends on that. I guess that I could say something like: 
```suggestion
There's not much in those but as you can see, the `Gemfile` has information on where to get the gems from and what gems are required. The `"~> 1.1"` is something called a version constraint, particularly one called pessimistic constraint. It relies on semantic versioning - in short, it is supposed to let your gem be updated but only until breaking changes have been introduced. It is equivalent to `gem "colorize", ">= 1.1", "<2.0"`.
```
and throw the rubygem's explanation in Additional Resources.